File: rust/mr/src/mr/worker.rs
use std::env::temp_dir;
use std::error::Error;
use std::fs::{self, File};
use std::io::{BufReader, Read, Write};

use crate::mapreduce::ReportArgs;
use crate::mapreduce::{map_reduce_client::MapReduceClient, RequestArgs, TaskType};
use mr_types::{KeyValue, MapFunc, ReduceFunc};
use tonic::{transport::Channel, Request};

fn ihash(key: &str) -> usize {
    key.as_bytes().iter().map(|&b| b as usize).sum()
}

// Handle map task. Return data to report to coordinator
fn handle_map(
    filename: String,
    task_no: usize,
    n_reduce: usize,
    mapf: MapFunc,
) -> Result<ReportArgs, Box<dyn Error>> {
    let mut f = File::open(&filename)?;
    let mut contents = String::new();
    f.read_to_string(&mut contents)?;

    unsafe {
        let mut kvs = mapf(filename.clone(), contents);
        kvs.sort_unstable();
        // println!("key-values: {:?}", kvs);

        let mut buckets: Vec<Vec<KeyValue>> = vec![vec![]; n_reduce];
        for kv in kvs {
            let k = ihash(&kv.key) % n_reduce;
            buckets[k].push(kv);
        }

        let mut intermediates = vec![];
        for i in 0..n_reduce {
            let fname = format!("mr-{}-{}", task_no, i);
            intermediates.push(fname.clone());

            let mut fp = File::create(&fname)?;

            let data = serde_json::to_string(&buckets[i])?;
            fp.write_all(data.as_bytes())?;
        }

        Ok(ReportArgs {
            kind: TaskType::Map as i32,
            task: filename,
            intermediates,
            ..Default::default()
        })
    }
}

// Read key-value pairs from intermediate files created by a Map task
fn read_intermediates(files: &Vec<String>) -> Result<Vec<KeyValue>, Box<dyn Error>> {
    let mut res = vec![];
    for fname in files {
        let file = File::open(fname)?;
        let reader = BufReader::new(file);
        let mut data: Vec<KeyValue> = serde_json::from_reader(reader)?;
        res.append(&mut data);
    }
    Ok(res)
}

// Handle reduce task.
fn handle_reduce(
    task_no: usize,
    files: &Vec<String>,
    reducef: ReduceFunc, //&dyn Fn(String, Vec<String>) -> String,
) -> Result<ReportArgs, Box<dyn Error>> {
    let mut kvs = read_intermediates(files)?;
    kvs.sort_unstable();

    let fname = format!("mr-out-{}", task_no);
    let tmpname = temp_dir().join(&fname);
    let mut tmp = File::create(&tmpname)?;

    // Call reduce on each distinct key and its associated values
    let mut i = 0;
    while i < kvs.len() {
        let (key, mut vals) = (kvs[i].key.clone(), vec![]);
        while i < kvs.len() && kvs[i].key == key {
            vals.push(kvs[i].value.clone());
            i += 1;
        }
        unsafe {
            tmp.write_all(format!("{} {}\n", key, reducef(key.clone(), vals)).as_bytes())?;
        }
    }

    fs::rename(tmpname, fname)?;

    Ok(ReportArgs {
        kind: TaskType::Reduce as i32,
        id: task_no as i32,
        ..Default::default()
    })
}

pub async fn work(
    w: &mut MapReduceClient<Channel>,
    mapf: MapFunc,
    reducef: ReduceFunc,
) -> Result<(), Box<dyn std::error::Error>> {
    loop {
        if false {
            break;
        }
        if let Ok(reply) = w.request_task(Request::new(RequestArgs::default())).await {
            let reply = reply.get_ref();
            if reply.done {
                break;
            }

            match reply.kind.try_into() {
                Ok(TaskType::Map) => {
                    println!("received map task: {}", reply.filename);
                    let args = handle_map(
                        reply.filename.clone(),
                        reply.task_no as usize,
                        reply.n_reduce as usize,
                        mapf,
                    )?;
                    w.report_task(Request::new(args)).await?;
                }
                Ok(TaskType::Reduce) => {
                    println!("received reduce task: {}", reply.task_no);
                    let args =
                        handle_reduce(reply.task_no as usize, &reply.intermediates, reducef)?;
                    w.report_task(Request::new(args)).await?;
                }
                _ => std::thread::sleep(std::time::Duration::from_secs(1)),
            }
        } else {
            break;
        }
    }

    Ok(())
}

File: go/mr/worker.go
package mr

import (
	"context"
	"encoding/json"
	"fmt"
	"hash/fnv"
	"io/ioutil"
	"log"
	"net"
	"os"
	"sort"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

// Map functions return a slice of KeyValue.
type KeyValue struct {
	Key   string
	Value string
}

type ByKey []KeyValue

func (a ByKey) Len() int           { return len(a) }
func (a ByKey) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByKey) Less(i, j int) bool { return a[i].Key < a[j].Key }

// use ihash(key) % NReduce to choose the reduce
// task number for each KeyValue emitted by Map.
func ihash(key string) int {
	h := fnv.New32a()
	h.Write([]byte(key))
	return int(h.Sum32() & 0x7fffffff)
}

// Request new task from Coordinator
func callRequestTask(w MapReduceClient) *RequestReply {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	reply, err := w.RequestTask(ctx, &RequestArgs{})
	if err != nil {
		reply.Done = true
	}
	return reply
}

// Report Map/Reduce results to Coordinator
func callReportTask(w MapReduceClient, args *ReportArgs) {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	_, err := w.ReportTask(ctx, args)
	if err != nil {
		log.Println(err)
	}
}

// Handle Map tasks, report intermediate files to Coordinator
func handleMap(w MapReduceClient, filename string, taskno int, nreduce int, mapf func(string, string) []KeyValue) {
	f, err := os.Open(filename)
	if err != nil {
		log.Fatalf("cannot open %v\n", filename)
	}
	text, err := ioutil.ReadAll(f)
	if err != nil {
		log.Fatalf("cannot read %v\n", filename)
	}
	f.Close()

	kvs := mapf(filename, string(text))
	sort.Sort(ByKey(kvs))

	buckets := make([][]KeyValue, nreduce)
	for _, kv := range kvs {
		key := ihash(kv.Key) % nreduce
		buckets[key] = append(buckets[key], kv)
	}

	intermediates := make([]string, nreduce)
	for i := 0; i < nreduce; i++ {
		fname := fmt.Sprintf("mr-%v-%v", taskno, i)
		intermediates[i] = fname
		f, _ := os.Create(fname)

		enc := json.NewEncoder(f)
		for _, kv := range buckets[i] {
			err := enc.Encode(&kv)
			if err != nil {
				log.Printf("Marshal error: %v\n", err)
			}
		}
		f.Close()
	}

	callReportTask(w, &ReportArgs{Kind: TaskType_Map, Task: filename, Intermediates: intermediates})
}

// Read key-value pairs from intermediate files created by Map task
func readIntermediates(files []string) []KeyValue {
	kvs := []KeyValue{}
	for _, filename := range files {
		f, err := os.Open(filename)
		if err == nil {
			dec := json.NewDecoder(f)
			for {
				var kv KeyValue
				if err := dec.Decode(&kv); err != nil {
					break
				}
				kvs = append(kvs, kv)
			}
		} else {
			log.Println("Reducer failed to open file: ", err.Error())
		}
		f.Close()
	}
	return kvs
}

// Handle Reduce tasks and report to Coordinator
func handleReduce(w MapReduceClient, taskno int, files []string, reducef func(string, []string) string) {
	kvs := readIntermediates(files)
	sort.Sort(ByKey(kvs))

	fname := fmt.Sprintf("mr-out-%v", taskno)
	tmp, err := ioutil.TempFile(".", fname)
	if err != nil {
		log.Printf("Reducer failed to create tempfile: %v\n", tmp)
	}

	// Call Reduce on each distinct key and its associated values
	// with results into temp file
	for i := 0; i < len(kvs); i++ {
		key, vals := kvs[i].Key, []string{}
		for i < len(kvs) && kvs[i].Key == key {
			vals = append(vals, kvs[i].Value)
			i++
		}
		i--
		fmt.Fprintf(tmp, "%v %v\n", key, reducef(key, vals))
	}
	// Rename temp file to result when finished
	os.Rename(tmp.Name(), fname)

	callReportTask(w, &ReportArgs{Kind: TaskType_Reduce, Id: int32(taskno)})
}

// A Worker performs Map or Reduce tasks at the behest of the Coordinator until
// all tasks are completed
func Worker(mapf func(string, string) []KeyValue, reducef func(string, []string) string) {
	dialer := func(ctx context.Context, addr string) (net.Conn, error) {
		var d net.Dialer
		return d.DialContext(ctx, "unix", addr)
	}
	sockname := CoordinatorSock()
	conn, err := grpc.Dial(sockname,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithContextDialer(dialer),
	)
	if err != nil {
		log.Fatalf("grpc.Dial(%q) failed: %v\n", sockname, err)
	}
	defer conn.Close()

	worker := NewMapReduceClient(conn)
	for {
		reply := callRequestTask(worker)
		if reply.Done {
			break
		}
		switch reply.Kind {
		case TaskType_Map:
			log.Printf("received map task: %s\n", reply.Filename)
			handleMap(worker, reply.Filename, int(reply.TaskNo), int(reply.NReduce), mapf)
		case TaskType_Reduce:
			log.Printf("received reduce task: %v\n", reply.TaskNo)
			handleReduce(worker, int(reply.TaskNo), reply.Intermediates, reducef)
		default:
			time.Sleep(time.Second)
		}
	}
}
